frame/assets/src/lib.rs

```diff
@@ -569,45 +592,207 @@ pub mod pallet {
+		#[pallet::weight(T::WeightInfo::start_destroy())]
+		pub fn start_destroy(
+			origin: OriginFor<T>,
+			#[pallet::compact] id: T::AssetId,
+		) -> DispatchResult {
+			let maybe_check_owner = match T::ForceOrigin::try_origin(origin) {
+				Ok(_) => None,
+				Err(origin) => Some(ensure_signed(origin)?),
+			};
+			let _ = Asset::<T, I>::try_mutate_exists(
+				id,
+				|maybe_details| -> Result<(), DispatchError> {
+					let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+					if let Some(check_owner) = maybe_check_owner {
+						ensure!(details.owner == check_owner, Error::<T, I>::NoPermission);
+					}
+					ensure!(details.is_frozen, Error::<T, I>::BadWitness);
+					details.status = AssetStatus::Destroying;
+
+					Self::deposit_event(Event::DestructionStarted { asset_id: id });
+					Ok(())
+				},
+			)?;
+			Ok(())
+		}
+
+		/// Destroy all accounts associated with a given asset.
+		/// `destroy_accounts` should only be called after `start_destroy` has been called, and the
+		/// asset is in a `Destroying` state
+		///
+		/// Due to weight restrictions, this function may need to be called multiple
+		/// times to fully destroy all accounts. It will destroy `RemoveKeysLimit` accounts at a
+		/// time.
+		///
+		/// The origin must conform to `ForceOrigin` or must be Signed and the sender must be the
+		/// owner of the asset `id`.
+		///
+		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
+		///   asset.
+		///
+		/// Each call Emits the `Event::DestroyedAccounts` event.
+		#[pallet::weight(T::WeightInfo::destroy_accounts(T::RemoveKeysLimit::get()))]
+		pub fn destroy_accounts(
 			origin: OriginFor<T>,
 			#[pallet::compact] id: T::AssetId,
-			witness: DestroyWitness,
 		) -> DispatchResultWithPostInfo {
 			let maybe_check_owner = match T::ForceOrigin::try_origin(origin) {
 				Ok(_) => None,
 				Err(origin) => Some(ensure_signed(origin)?),
 			};
-			let details = Self::do_destroy(id, witness, maybe_check_owner)?;
-			Ok(Some(T::WeightInfo::destroy(
-				details.accounts.saturating_sub(details.sufficients),
-				details.sufficients,
-				details.approvals,
-			))
-			.into())
+			let mut dead_accounts: Vec<T::AccountId> = vec![];
+			let mut removed_accounts = 0;
+			let _ = Asset::<T, I>::try_mutate_exists(
+				id,
+				|maybe_details| -> Result<(), DispatchError> {
+					let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+					if let Some(check_owner) = maybe_check_owner {
+						ensure!(details.owner == check_owner, Error::<T, I>::NoPermission);
+					}
+					ensure!(details.is_frozen, Error::<T, I>::BadWitness);
+					// Should only destroy accounts while the asset is being destroyed
+					ensure!(details.status == AssetStatus::Destroying, Error::<T, I>::LiveAsset);
+
+					for (who, v) in Account::<T, I>::drain_prefix(id) {
+						let _ = Self::dead_account(&who, &mut details, &v.reason, true);
```

Comment:
you probably want the bench run only the heaviest case of this function

Reply:
Could you explain more? Do you mean I should benchmark only with the `RemoveKeyLimit `  and not random numbers between 0 and the limit?

Reply:
The amount of work the function does depends from the `reason` here.
For example If reason `DepositRefunded`, the function will only do 
https://github.com/paritytech/substrate/blob/10fdd9090b69ff6e42a49ef2c203ea080d92020b/frame/assets/src/functions.rs#L102

If reason `Consumer`, it mutates the storage item and does the same saturating_sub as above
https://github.com/paritytech/substrate/blob/10fdd9090b69ff6e42a49ef2c203ea080d92020b/frame/assets/src/functions.rs#L93

In the bench you want to record the heavies case (reason) of this function which is valid in this context (the context of destroy_accounts func).
Or ask a user of the function to let you know how many of each reason are there (witness data), but this might be too complex.

Reply:
In the benchmark, I focus on sufficients. Does that cover this case?
```rust

	destroy_accounts {
		let c in 0 .. T::RemoveKeysLimit::get();
		let (caller, _) = create_default_asset::<T, I>(true);
		add_sufficients::<T, I>(caller.clone(), c);
		Assets::<T, I>::freeze_asset(
			SystemOrigin::Signed(caller.clone()).into(),
			Default::default(),
		)?;
		Assets::<T,I>::start_destroy(SystemOrigin::Signed(caller.clone()).into(), Default::default());
	}:_(SystemOrigin::Signed(caller), Default::default())
	verify {
		assert_last_event::<T, I>(Event::DestroyedAccounts {
			asset_id: Default::default() ,
			accounts_destroyed: c,
			accounts_remaining: 0,
		}.into());
	}
```